library/math/num.lisp

```diff
@@ -161,6 +161,35 @@
 ;;; Num instances for integers
 ;;;
 
+(cl:eval-when (:compile-toplevel :load-toplevel)
+  (cl:defmacro define-unsigned-num-underflow (type bits)
+    "Define a `Num' instance for Type which signals an error on underflow."
+    `(define-instance (Num ,type)
+       (define (+ a b)
+         (lisp ,type (a b)
+	   (cl:if (cl:and (cl:< b 0) (cl:< a (cl:- 0 b)))
+		  (cl:error ,(cl:format cl:nil "Unsigned value underflowed ~D bits." bits))
+		  (cl:+ a b))))
+       
+       (define (- a b)
+         (lisp ,type (a b)
+	   (cl:if (cl:and (cl:>= b 0) (cl:< a (cl:+ 0 b)))
+		  (cl:error ,(cl:format cl:nil "Unsigned value underflowed ~D bits." bits))
+		  (cl:- a b))))
+       
+       (define (* a b)
+	 (lisp ,type (a b)
+	   (cl:if (cl:or (cl:and (cl:and (cl:> b 0) (cl:< a 0)) (cl:< a (cl:/ 0 b)))
```

Comment:
unsigned multiplication can't underflow

Reply:
what do you mean by this? are you saying the code that i wrote doesnt detect the underflow or that in general unsigned multiplication doesnt underflow?

wouldnt this be an underflow: `2 * -3`?

Reply:
-3 isn't an unsigned number